Lib/pdb.py

```diff
@@ -624,11 +626,97 @@ def _disable_command_completion(self):
             self.completenames = completenames
         return
 
+    def _exec_in_closure(self, source, globals, locals):
+        """ Run source code in closure so code object created within source
+            can find variables in locals correctly
+
+            returns True if the source is executed, False otherwise
+        """
+
+        # Determine if the source should be executed in closure. Only when the
+        # source compiled to multiple code objects, we should use this feature.
+        # Otherwise, we can just raise an exception and normal exec will be used.
+
+        code = compile(source, "<string>", "exec")
+        for const in code.co_consts:
+            if isinstance(const, CodeType):
+                break
+        else:
+            return False
+
+        # locals could be a proxy which does not support pop
+        # copy it first to avoid modifying the original locals
+        locals_copy = dict(locals)
+
+        locals_copy["__pdb_eval__"] = {
+            "result": None,
+            "write_back": {}
+        }
+
+        # If the source is an expression, we need to print its value
+        try:
+            compile(source, "<string>", "eval")
+        except SyntaxError:
+            pass
+        else:
+            source = "__pdb_eval__['result'] = " + source
+
+        # Add write-back to update the locals
+        source = ("try:\n" +
+                  textwrap.indent(source, "  ") + "\n" +
+                  "finally:\n" +
+                  "  __pdb_eval__['write_back'].update(locals())")
+
+        local_vars = list(locals_copy.keys())
+
+        # Build a closure source code with freevars from locals like:
+        # def outer():
+        #   var = None
+        #   def __pdb_scope():  # This is the code object we want to execute
+        #     nonlocal var
+        #     <source>
+        #   return __pdb_scope.__code__
+        source_with_closure = ("def __pdb_outer():\n" +
+                               "\n".join(f"  {var} = None" for var in local_vars) + "\n" +
+                               "  def __pdb_scope():\n" +
+                               "\n".join(f"    nonlocal {var}" for var in local_vars) + "\n" +
+                               textwrap.indent(source, "    ") + "\n" +
+                               "  return __pdb_scope.__code__"
+                               )
+
+        # Get the code object of __pdb_scope()
+        # The exec fills locals_copy with the __pdb_outer() function and we can call
+        # that to get the code object of __pdb_scope()
+        code = exec(source_with_closure, {}, locals_copy)
+        code = locals_copy.pop("__pdb_outer")()
+
+        cells = tuple(types.CellType(locals_copy.get(var)) for var in code.co_freevars)
+
+        try:
+            exec(code, globals, locals_copy, closure=cells)
+        except Exception:
+            return False
```

Comment:
If this raises, we should just let it propagate, right? Otherwise we fall back on `exec`'ing it normally which could change the exception? Or are there cases where this could fail and the correct evaluation isn't in a closure?

```suggestion
        exec(code, globals, locals_copy, closure=cells)
```

Reply:
And if this is true, it's probably cleaner to separate this function into two: one for deciding if we need to run in a closure, and another for doing it.

Reply:
This is intentional - any exception in the new method, we fall back to the old pure `exec`. There are two reasons behind it:
1. To keep maximum backwards compatibility, making sure this is an incremental improvement.
2. Because of the extra code secretly added, some of the user code may conflict with it and they will have no idea when they saw the exception. For example, if they do somethign like `global x`, which conflicts with `nonlocal x`.

Reply:
> 2\. For example, if they do somethign like `global x`, which conflicts with `nonlocal x`.

But that would fail earlier, when `exec`'ing `source_with_closure`, right? Are there other cases where this can raise incorrectly?

What I'm concerned about is if the code the user's running has a bug, and instead of raising the "correct" error here, we swallow it, and give them an "incorrect" error (or worse, silently succeed) by falling back on a non-closure `exec`.

If you're confident that can't happen, though, then this is fine.

Reply:
I can't think of good examples here, but in my mind, the users always get the "wrong" error for the code that supposed to have the "correct" error - we didn't change that. Without any concrete example, I'd like to keep things the same in the dark area - do not introduce any new errors to code that's working before, leave the errors that we could report better as it is for now.

An extreme yet not that convincing example would be if the user uses variable like `__pdb_eval__` in their code - that's some artifact we introduced to make this happen, not part of the original code.

If in the future we had a real case where keeping the exception of our new method is helpful, we can change it.

Reply:
Okay, makes sense!